File: packages/mobile_ios/App/PairingModel.swift
import Foundation
import Observation
import Security

/// Routes and formats input only. The core validates links, owns retries/nonces
/// after submission, proposes trust, and commits credentials through Keychain.
enum PairingInput {
    static func routes(_ url: URL) -> Bool {
        let raw = url.absoluteString
        return raw.hasPrefix("verde://pair?") || raw.hasPrefix("https://verdeai.dev/pair?")
    }

    static func manual(host: String, grant: String, code: String) -> String {
        var link = URLComponents()
        link.scheme = "verde"
        link.host = "pair"
        link.queryItems = [URLQueryItem(name: "host", value: host), URLQueryItem(name: "grant_id", value: grant)]
        link.fragment = "code=\(code)"
        return link.string ?? ""
    }

    static func nonce() throws -> String {
        var bytes = [UInt8](repeating: 0, count: 16)
        guard SecRandomCopyBytes(kSecRandomDefault, bytes.count, &bytes) == errSecSuccess else {
            throw CoreBridgeError.invalidOutput
        }
        return bytes.map { String(format: "%02x", $0) }.joined()
    }
}

/// One saved host's core session. The platform owns lifecycle/network signals
/// and pairing input; the core owns every protocol decision.
@MainActor @Observable
final class PairingModel {
    let store: CoreViewStore
    private(set) var host: CoreHost?
    private let makeHost: (CoreViewStore) throws -> CoreHost
    private var opening: Task<Void, Never>?
    private var started = false
    private var openFailed = false
    // Lifecycle/network events reach the core in call order, after `start`, deduplicated.
    private var signals: [Event] = []
    private var draining = false
    private var lastForeground: Bool?
    private var lastNetwork: NetworkState?
    private var pendingLink: String?
    private(set) var intentID: String?
    private(set) var submitting = false
    private(set) var inputError: String?
    var deviceLabel: String

    init(deviceLabel: String, store: CoreViewStore? = nil,
         makeHost: @escaping (CoreViewStore) throws -> CoreHost) {
        self.deviceLabel = deviceLabel
        self.store = store ?? CoreViewStore()
        self.makeHost = makeHost
    }

    var row: HostView? { store.hosts?.data?.items.first }
    var operation: Operation? { store.hosts?.data?.operations.first { $0.intent_id == intentID } }
    var busy: Bool { submitting || pendingLink != nil || (intentID != nil && operation == nil) || operation?.state == "pending" }
    var paired: Bool { row?.auth_state == "paired" && operation?.state != "pending" }
    /// Paired with an accepted identity: a new link opens another host slot instead.
    var complete: Bool { row?.auth_state == "paired" && row?.trust_proposal == nil }
    var fatal: Bool { store.failed || openFailed }
    /// A received link waits for the core's stored state before it is submitted.
    var awaitingSubmit: Bool { pendingLink != nil && !submitting }
    var error: LocalError? { row?.error ?? operation?.error }
    var errorText: String? {
        if let inputError { return inputError }
        if fatal { return "The connection stopped unexpectedly. Close and reopen Verde to try again." }
        guard let error else { return nil }
        if error.domain == "storage" { return "Could not save or load this host in Keychain. Unlock your phone, then retry." }
        switch error.code {
        case "trust_denied": return "Host was not trusted. Scan a new code when you are ready."
        case "auth_rejected": return "This pairing grant expired or was already used. Create a new grant on your host."
        case "exchange_uncertain": return "The host may have paired this phone, but its reply was lost. Create a new grant."
        case "repair_required": return "This device's access was revoked. Pair again with a new grant."
        case "tls_rejected": return "The host's secure connection could not be verified. Check its certificate and address."
        case "network_unavailable": return "Could not reach the host. Check Tailscale and your network, then retry."
        case "discovery_rejected", "identity_rejected": return "The host identity could not be verified. Check the host before pairing again."
        default: return "Pairing could not complete. Check that Verde is up to date on your phone and host."
        }
    }

    /// Opens the core once. It starts in the background; queued signals then
    /// follow in order (network before foreground when both are known).
    func start() async {
        if let opening { await opening.value; return }
        let task = Task { await openCore() }
        opening = task
        await task.value
    }

    private func openCore() async {
        do {
            let host = try makeHost(store)
            self.host = host
            try await host.send(.start(EventStart(now_ms: 0, wall_time_ms: 0, foreground: false, network_available: true)))
            started = true
            await drain()
        } catch {
            openFailed = true
            inputError = "Could not open the connection. Close and reopen Verde to try again."
            store.onApply?()
        }
    }

    func foreground(_ active: Bool) {
        guard lastForeground != active else { return }
        lastForeground = active
        enqueue(active ? .foreground(EventForeground(now_ms: 0, wall_time_ms: 0))
            : .background(EventBackground(now_ms: 0, wall_time_ms: 0)))
    }

    /// The core invalidates transport on a changed route and reconnects with its jittered backoff.
    func network(_ state: NetworkState) {
        guard lastNetwork != state else { return }
        lastNetwork = state
        enqueue(.network_changed(EventNetworkChanged(now_ms: 0, wall_time_ms: 0,
            available: state.available, network_id: state.id)))
    }

    private func enqueue(_ event: Event) {
        signals.append(event)
        Task { await drain() }
    }

    private func drain() async {
        guard started, !draining, let host else { return }
        draining = true
        defer { draining = false }
        while !signals.isEmpty {
            let event = signals.removeFirst()
            do { try await host.send(event) }
            catch CoreBridgeError.rejected { continue } // No state change; the next signal still applies.
            catch { signals.removeAll(); return }         // The host closed and published its failure.
        }
    }

    func receive(_ link: String) async {
        guard !busy else { return }
        guard link.utf8.count <= 8192 else { inputError = "This pairing link is too long."; return }
        pendingLink = link.trimmingCharacters(in: .whitespacesAndNewlines)
        inputError = nil
        await start()
        await submitPending()
    }

    func open(_ url: URL) async {
        guard PairingInput.routes(url) else { inputError = "This is not a Verde pairing link."; return }
        await receive(url.absoluteString)
    }

    func submitPending() async {
        guard !submitting, let host, let link = pendingLink,
              let row, row.auth_state != "loading", row.lifecycle == .foreground else { return }
        pendingLink = nil
        submitting = true
        defer { submitting = false }
        let id = UUID().uuidString
        do {
            let nonce = try PairingInput.nonce()
            try await host.send(.pair(EventPair(now_ms: 0, wall_time_ms: 0, intent_id: id,
                link: link, device_label: deviceLabel, client_nonce: nonce)))
            intentID = id
        } catch {
            inputError = "Invalid pairing details. Use the complete link or check the host, grant ID, code and device name."
        }
    }

    func trust(_ proposal: TrustProposal, accept: Bool) async {
        guard !submitting else { return }
        await send(.trust_decision(EventTrustDecision(now_ms: 0, wall_time_ms: 0,
            intent_id: UUID().uuidString, proposal_id: proposal.id, accept: accept)))
    }

    func retry() async {
        inputError = nil
        await send(.retry_connection(EventRetryConnection(now_ms: 0, wall_time_ms: 0, intent_id: UUID().uuidString)))
    }

    private func send(_ event: Event) async {
        guard let host else { return }
        submitting = true
        defer { submitting = false }
        do { try await host.send(event) }
        catch { inputError = "This action is no longer available. Check the connection and try again." }
    }

    func clearNotice() { inputError = nil }

    func stop() async {
        pendingLink = nil
        signals.removeAll()
        await opening?.value
        try? await host?.shutdown()
        host = nil
    }
}

File: packages/mobile_android/app/src/main/java/dev/verdeai/app/PairingModel.kt
package dev.verdeai.app

import android.app.Application
import android.os.Build
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.ViewModel
import dev.verdeai.core.*
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.*
import java.net.URLEncoder
import java.security.SecureRandom
import java.util.UUID

class VerdeApplication : Application() {
    val secureStore by lazy { AndroidSecureStore(this) }
    val viewCache by lazy { AndroidSecureStore(this, "view-cache") }
    internal val signals: AppSignals by lazy { AndroidAppSignals(this) }
    /** Process-scoped so rotation and Activity recreation never relock; gates UI only. */
    internal val appLock by lazy { AppLockModel(secureStore, signals.foreground, { deviceSecure(this) }) }
}

internal data class PairingState(
    val host: HostView? = null,
    val operation: Operation? = null,
    val busy: Boolean = false,
    val notice: String? = null,
    val fatal: Boolean = false,
) {
    val error get() = operation?.error ?: host?.error
    val canEnter get() = !fatal && !busy && host != null && host.lifecycle == Lifecycle.foreground &&
        host.auth_state != "loading" && host.auth_state != "paired" &&
        host.trust_proposal == null && operation?.state != "pending" && error?.domain != "storage"
    val complete get() = host?.auth_state == "paired" && host.trust_proposal == null
}

/** UI state only. The core retains the grant/nonce and decides all retry and trust policy. */
internal class PairingModel(private val createHost: suspend () -> CoreHost) : ViewModel() {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val mutableState = MutableStateFlow(PairingState())
    val state = mutableState.asStateFlow()
    // Deliberately not SavedStateHandle/rememberSaveable: pair material dies with the process.
    var link by mutableStateOf("")
    var manualHost by mutableStateOf("")
    var grant by mutableStateOf("")
    var code by mutableStateOf("")
    var deviceLabel by mutableStateOf(Build.MODEL)
    private var pairIntent: String? = null
    // Lifecycle/network events are delivered in call order, after `start`, and deduplicated.
    private val signals = Channel<(Long, Long) -> Event>(Channel.UNLIMITED)
    private var lastForeground: Boolean? = null
    private var lastNetwork: NetworkState? = null
    private val host = scope.async { createHost().also {
        try { it.send { n,w -> EventStart(now_ms=n, wall_time_ms=w, foreground=false, network_available=true) } }
        catch (e: Exception) { it.close(); throw e }
    } }

    init {
        scope.launch {
            try {
                val core = host.await()
                launch { core.hosts.collect { query ->
                    mutableState.update { it.copy(host=query?.data?.items?.firstOrNull(),
                        operation=query?.data?.operations?.find { op -> op.intent_id == pairIntent } ?: it.operation) }
                } }
                launch { core.failed.collect { failed -> if (failed) mutableState.update { it.copy(fatal=true) } } }
                for (signal in signals) {
                    try { core.send(signal) }
                    catch (_: CoreInputRejected) { }
                }
            } catch (e: CancellationException) { throw e }
            catch (_: Exception) { mutableState.update { it.copy(fatal=true) } }
        }
    }

    fun receiveLink(value: String) {
        // Receiving an Intent or QR never authorizes exchange or replaces an in-flight grant.
        if (state.value.busy || state.value.operation?.state == "pending" || state.value.complete) {
            notice("Finish the current pairing before opening another link.")
        } else if (value.length > 8192) notice("This pairing link is too long. Create a new link on the host.")
        else { link = value; notice(null) }
    }

    fun pair(manual: Boolean = false) {
        if (!state.value.canEnter) return
        val value = if (manual) manualPairLink(manualHost.trim(), grant.trim(), code.trim()) else link.trim()
        val label = deviceLabel.trim()
        if (value.isEmpty() || label.isEmpty()) { notice("Enter a pairing link and a device label."); return }
        val id = UUID.randomUUID().toString()
        val nonce = ByteArray(16).also(SecureRandom()::nextBytes).joinToString("") { "%02x".format(it) }
        pairIntent = id
        mutableState.update { it.copy(operation=Operation(id, "pending", null)) }
        clearInputs()
        action(pairSubmission=true) { core -> core.send { n,w -> EventPair(now_ms=n, wall_time_ms=w, intent_id=id,
            link=value, device_label=label, client_nonce=nonce) } }
    }

    fun trust(proposalId: String, accept: Boolean) = action { core ->
        core.send { n,w -> EventTrustDecision(now_ms=n, wall_time_ms=w,
            intent_id=UUID.randomUUID().toString(), proposal_id=proposalId, accept=accept) }
    }
    fun retry() = action { core -> core.send { n,w -> EventRetryConnection(now_ms=n,
        wall_time_ms=w, intent_id=UUID.randomUUID().toString()) } }
    fun foreground(active: Boolean) {
        if (lastForeground == active) return
        lastForeground = active
        signals.trySend { n,w -> if (active) EventForeground(now_ms=n, wall_time_ms=w) else EventBackground(now_ms=n, wall_time_ms=w) }
    }
    /** The core invalidates transport on a changed route and reconnects with its jittered backoff. */
    fun network(state: NetworkState) {
        if (lastNetwork == state) return
        lastNetwork = state
        signals.trySend { n,w -> EventNetworkChanged(now_ms=n, wall_time_ms=w, available=state.available, network_id=state.id) }
    }
    fun notice(text: String?) { mutableState.update { it.copy(notice=text) } }
    private fun clearInputs() { link=""; manualHost=""; grant=""; code="" }
    private fun action(pairSubmission: Boolean = false, block: suspend (CoreHost) -> Unit) {
        if (state.value.busy || state.value.fatal) return
        mutableState.update { it.copy(busy=true, notice=null) }
        scope.launch {
            try { block(host.await()) }
            catch (e: CoreInputRejected) {
                if (pairSubmission) mutableState.update { it.copy(operation=null) }
                notice(if (e.status == 1) "Invalid pairing link or device label. Copy a fresh link from Verde."
                    else "The host is not ready for this action. Try again after it finishes loading.")
            } catch (_: Exception) { mutableState.update { it.copy(fatal=true) } }
            finally { mutableState.update { it.copy(busy=false) } }
        }
    }
    internal suspend fun coreHost(): CoreHost = host.await()

    override fun onCleared() = dispose()
    internal fun dispose() {
        clearInputs()
        signals.close()
        // Own cleanup beyond scope cancellation; CoreHost closes its dispatcher/executor.
        scope.launch {
            try { host.await().close() } catch (_: Exception) { }
            finally { scope.cancel() }
        }
    }
}

internal fun manualPairLink(host: String, grant: String, code: String): String {
    fun encode(value: String) = URLEncoder.encode(value, "UTF-8").replace("+", "%20")
    return "verde://pair?host=${encode(host)}&grant_id=${encode(grant)}#code=${encode(code)}"
}

internal fun pairingError(error: LocalError): String = when {
    error.domain == "storage" -> "Could not save or read the secure credential. Unlock your phone and retry."
    error.code == "grant_rejected" -> "This pairing grant has expired, was already used, or was rejected. Create a new pairing link on the host."
    error.code == "auth_rejected" -> "The host rejected authentication. Check its paired-device settings."
    error.code == "network_unavailable" || error.failure_kind == "network" -> "Host unreachable. Is Tailscale on? Check that your phone and host are connected, then retry."
    error.code == "exchange_uncertain" -> "The host may have paired this phone, but no response arrived. Create a new grant; this exchange cannot be safely retried."
    error.code == "trust_denied" -> "Host not trusted. No pairing credential was sent."
    error.code == "tls_rejected" -> "The host's secure connection could not be verified. Check its certificate before trying a new pairing link."
    error.code == "repair_required" -> "This device needs to pair again. Create a new pairing link on the host."
    error.code == "protocol_rejected" -> "This host is not compatible. Update Verde on the host and phone."
    else -> "Pairing could not complete. Check the host and try again."
}
